website/blog/2019-07-15-complete-table-scan.md

```diff
@@ -0,0 +1,169 @@
+
+The initial impulse motivating this work is the observation that table scan is by far the #1 operator in Presto workloads I have seen. This is a little over half of all Presto CPU, with repartitioning a distant second, at around 1/10 of the total. The other half of the motivation is ready opportunity: Presto in its pre-Aria state does almost none of the things that are common in table scan.
+
+For easy reproducibility and staying away from proprietary material, we use a TPC-H 100G dataset running on a desktop machine with 2x4 Skylake cores at 3.5GHz. The data is compressed with Snappy and we are running with warm OS cache. The Presto is a modified 0.221 where the Aria functionality can be switched on and off. Not to worry, we will talk about disaggregated storage and IO in due time but the basics will come first.
+
+# Simple scan
+
+The base case for scan optimization is the simplest possible query:
+
+```sql
+SELECT SUM(l_extendedprice)
+FROM lineitem
+WHERE suppkey = 12345;
+```
+| Version  | Wall time (seconds) | CPU time (seconds) |
+| -------- | ------------------- | ------------------ |
+| Aria     | 2                   | 14.2               |
+| Baseline | 3                   | 29.0               |
+
+
+This is a 2x win for the most basic case. This selects one row of a million and spends all its time in scan. It is here worthwhile to step through exactly how each implementation does this, as we need to understand this when analyzing the more complex cases.
+
+## Mechanics of a scan
+
+Baseline Presto does this as follows: The scan `OrcPageSource` produces consecutive `Page` instances that contain a `LazyBlock` for each column. This operation as such takes no time since the `LazyBlock` instances are just promises. The actual work takes place when evaluating the generated code for the comparison. This sees that the column is not loaded, loads all the values in the range of the `LazyBlock`, typically 1024 values and then does the operation and produces a set of passing row numbers. This set is empty for all but 1/1000 of the cases. If this is empty, the `LazyBlock` for `l_extendedprice` is not touched. If there are hits, the `l_extendedprice` `LazyBlock` is loaded and the values for the selected rows are copied out. When this happens, 1024 values are decoded from the column and most often one of them is accessed. Loading a `LazyBlock` allocates memory for each value. In the present case this becomes garbage immediately after first use. The same applies to the values in extended price, of which only one is copied to a `Block` of output. This is handled by a special buffering stage that accumulates rows from multiple loaded `LazyBlock` instances until there is a minimum batch worth of rows to pass to the next operator.
```

Comment:
- 1/1000 may need to be revised to match the example above - 1/100K
- l_extendedprice -> extendedprice

Reply:
Changed to 1/100k.